File: Task/Erd-s-primes/Julia/erd-s-primes.jl
using Primes, Formatting

function isErdős(p::Integer)
    isprime(p) || return false
    for i in 1:100
        kfac = factorial(i)
        kfac >= p && break
        isprime(p - kfac) && return false
    end
    return true
end

const Erdőslist = filter(isErdős, 1:1000000)
const E2500 = filter(x -> x < 2500, Erdőslist)

println(length(E2500), " Erdős primes < 2500: ", E2500)
println("The 7875th Erdős prime is ", format(Erdőslist[7875], commas=true))

File: Task/Erd-s-primes/Nim/erd-s-primes.nim
import math, sets, strutils, sugar

const N = 1_000_000

# Sieve of Erathostenes.
var isComposite: array[2..N, bool]
for n in 2..N:
  let n2 = n * n
  if n2 > N: break
  if not isComposite[n]:
    for k in countup(n2, N, n):
      isComposite[k] = true

template isPrime(n: int): bool = n > 1 and not isComposite[n]

let primeList = collect(newSeq):
                  for n in 2..N:
                    if n.isPrime: n

const Factorials = collect(newSeq):
                     for n in 1..20:
                       if fac(n) >= N: break
                       fac(n)


proc isErdösPrime(p: int): bool =
  ## Check if prime "p" is an Erdös prime.
  for f in Factorials:
    if f >= p: break
    if (p - f).isPrime: return false
  result = true


let erdösList2500 = collect(newSeq):
                      for p in primeList:
                        if p >= 2500: break
                        if p.isErdösPrime: p

echo "Found $# Erdös primes less than 2500:".format(erdösList2500.len)
for i, prime in erdösList2500:
  stdout.write ($prime).align(5)
  stdout.write if (i+1) mod 10 == 0: '\n' else: ' '
echo()

var erdös7875: int
var count = 0
for p in primeList:
  if p.isErdösPrime: inc count
  if count == 7875:
    erdös7875 = p
    break
echo "\nThe 7875th Erdös prime is $#.".format(erdös7875)
